Telink b91 tlsr9518adk serial gpio

yurvyn opened the pull request:
Basic support for Telink B91 SoC and TLSR9518ADK board:
- PinMux Driver
- Serial Driver
- GPIO Driver

galak commented on drivers/gpio/Kconfig.b91:
```diff
@@ -0,0 +1,10 @@
+
+# Telink B91 GPIO configuration options
+
+config GPIO_B91
```
Probably good to rename this to GPIO_TELINK_B91

yurvyn replied:
Done

galak commented on drivers/pinmux/Kconfig.b91:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2021 Telink Semiconductor
+# SPDX-License-Identifier: Apache-2.0
+
+config PINMUX_B91
```
rename to PINMUX_TELINK_B91

yurvyn replied:
Done

galak commented on drivers/serial/Kconfig.b91:
```diff
@@ -0,0 +1,12 @@
+
+# Telink B91 UART configuration options
+
+config UART_B91
```
rename to UART_TELINK_B91

yurvyn replied:
Done

galak commented on soc/riscv/riscv-privilege/telink_b91/Kconfig.soc:
```diff
@@ -0,0 +1,39 @@
+	default y
+	depends on SOC_SERIES_RISCV_TELINK_B91
+
+config HWDSP
```
What exactly is HWDSP?  Is this unique to TELINK?  Maybe rename the symbol if so.

yurvyn replied:
HWDSP is specific for Telink B91 (NDS32) Toolchain, it enabled DSP commands in compiler.
Done, renamed to TELINK_B91_HWDP.

galak commented on boards/riscv/tlsr9518adk80d/tlsr9518adk80d.yaml:
```diff
@@ -0,0 +1,10 @@
+type: mcu
+arch: riscv32
+toolchain:
+  - cross-compile
```
Can we not use the zephyr toolchain to build this?

yurvyn replied:
To build B91 solution Telink specific (NDS32) toolchain is required.
Currently this toolchain is not integrated to SDK-NG project, users need to download this toolchain from Telink website.
The board documentation contains info how to gat the toolchain and how to build.

galak replied:
What is unique or different in the telink toolchain?  Is it possible to build a basic version utilizing the common RISC-V support in the Zephyr SDK?  I understand that it may not expose all the features of the hardware, but at least we'd be able to get CI coverage for the platform.

yurvyn replied:
> What is unique or different in the telink toolchain?

I do not know what exactly the difference is. I will investigate it and discuss with my colleagues when I come back from vacation in a week, on May 31.

> Is it possible to build a basic version utilizing the common RISC-V support in the Zephyr SDK? I understand that it may not expose all the features of the hardware, but at least we'd be able to get CI coverage for the platform.

No, currently it is not possible. Zephyr SDK contains only RISC-V 64 bits toolchain version, Telink B91 is RISC-V 32.
We already have a task in our backlog to investigate the integration of our toolchain to Zephyr SDK-NG project.
However, we are planning to do this investigation as a second step, the first step is to contribute B91 solution to Zephyr as cross-compile.

carlescufi replied:
>No, currently it is not possible. Zephyr SDK contains only RISC-V 64 bits toolchain version, Telink B91 is RISC-V 32.

I am pretty sure we build RV32 platforms with the RV64 toolchain already today, for example [this one](https://github.com/zephyrproject-rtos/zephyr/tree/main/soc/riscv/openisa_rv32m1). Also, #34970 seems to be able to use the Zephyr SDK toolchain to build RV32 images for the GigaDevice SoCs. Could you please find out if there is info we are missing here?

yurvyn replied:
I have just checked usage of RISC-V toolchain from zephyr-sdk and the solution is not compilable.
Some registers (which we are using in B91 SoC "start.s" and "soc_irq.s" files) are unknown for this toolchain (mcache_ctl, mmisc_ctl, mxstatus). Also, "hal_telink/tlsr9/drivers/b91/core.h" includes "nds_intrinsic.h" which is part of NDS32 toolchain.

Looks like it is not possible to use the generic RISC-V toolchain since Telink B91 SoC has NDS32 core, which actually is a new architecture based on RISC-V provided by Andes Technology corporation.

@carlescufi, @galak
I propose to release B91 solution as cross-compile first and only after that integrate NDS32 RISC-V toolchain to zephyr-sdk.
Are you OK with that?

carlescufi replied:
Wait, but then this is not a RISC-V core, is it? It's actually a new architecture, NDS32? Because I see that [Linux](https://elixir.bootlin.com/linux/latest/source/arch/nds32) has its own arch for it. In this case, we need to add a new arch in `arch/`? CC @stephanosio and @tejlmand

carlescufi replied:
OK, I see this in the Andes website: 

>AndesCore processors are based on the AndeStar™ V5 Instruction Set Architecture (ISA), which is compliant to the RISC-V technology.

So I assume that this is, as you mentioned, a customized RISC-V processor. But the NDS32 architecture seems supported by [GCC](https://gcc.gnu.org/backends.html), so I am a bit confused. Adding a requirement for an external toolchain is quite a heavy one, so I'd very much like to get to the bottom of this before we move forward.

stephanosio replied:
> Telink B91 SoC has NDS32 core, which actually is a new architecture based on RISC-V

A quick search suggests that "NDS32" refers to the "AndeStar architecture" which was initially based on Andes's own ISA, but later adopted the RISC-V ISA (see http://www.andestech.com/en/products-solutions/andestar-architecture/).

@yurvyn What version of "AndeStar architecture" is TLSR9 based on?

stephanosio replied:
[So it seems TLSR9 uses Andes D25F core](https://www.cnx-software.com/2020/11/04/telink-tlsr9-wireless-audio-iot-risc-v-soc-integrates-risc-v-dsp-simd-p-extension/
), which is AndeStar V5 (i.e. RISC-V + Andes extensions).

stephanosio replied:
So I've taken a look at the Telink TLSR9 toolchain and it is, without a doubt, binutils/gcc targeting `riscv32-elf`, which is NOT the `nds32-*` target (`nds32-elf` would be for the AndeStar V3 architecture, which is based on Andes's own ISA).

Though, it seems to be heavily patched to support Andes's extension instructions that are not part of the standard RISC-V spec.

See [`nds_intrinsic.h`](https://github.com/andestech/gcc/blob/bsp-v5_1_0/gcc/config/riscv/nds_intrinsic.h).

It looks like these patches have not been upstreamed (or likely there is no plan to do so) and therefore a separate toolchain (possibly `ndsriscv-zephyr-elf`) will be required to support TLSR9.

yurvyn replied:
Yes, all of the above is correct.

As I already mentioned, I propose to release B91 solution as cross-compile first and only after that investigate integration of NDS32 RISC-V toolchain to zephyr-sdk. Probably need to contact Andes Technology and ask them to add NDS32 toolchain patches to the upstream.

stephanosio replied:
> Probably need to contact Andes Technology and ask them to add NDS32 toolchain patches to the upstream.

That would be ideal, especially noting that many of the intrinsics in `nds_intrinsic.h` seem to be part of the upcoming RISC-V "P extension" which is not finalised yet.

carlescufi replied:
>As I already mentioned, I propose to release TLSR9 solution as cross-compile first and only after that investigate integration of NDS32 RISC-V toolchain to zephyr-sdk. Probably need to contact Andes Technology and ask them to add NDS32 toolchain patches to the upstream.

Given that this is currently pending standarization, this sounds like a reasonable compromise, but I'd like the Andes folks, and @galak to chime in as well.

cwshu replied:
@carlescufi Thank you very much for informing this, we haven't found this PR before.

Andes already has Zephyr porting of AndeStar v5 cores (e.g. Andes D25F), and it supports generic RISC-V toolchain if NOT using any V5 extension instructions. This is the [Zephyr source](https://github.com/cwshu/zephyr/tree/andes_v5_demo) of Andes V5 platform (build by `cmake -DBOARD=ae350`)
and entire kernel testcases (`tests/kernel`) can be compiled by generic RISC-V toolchain.

I think TLSR9 SoC could be based on it to support generic RISC-V toolchain (e.g. `soc_v5.h` for custom CSRs). We also plan to upstream Andes V5 platform (soft-cores on the FPGA) to Zephyr community recently.
(p.s. Andes porting is Zephyr v2.4 currently, but I will upgrade SoC/board porting to main branch quickly.)

Andes V5 platform already has 8 Zephyr drivers and has `CONFIG_NOCACHE_MEMORY` support by Andes implementation of RISC-V PMA. Also, we have RISC-V SMP support on Zephyr 2.4 [#29131] and is working on newer version Zephyr.

yurvyn replied:
@cwshu,
Thank you very much for the info and for pointing me to the Andes V5 branch.
In this branch I found some useful info about NDS32 arch and now I am able to use the toolchain from zephyr-sdk.

@carlescufi,
I need a few days to finalize the migration to zepkyr-sdk toolchain usage and to finish with the testing.
Will provide new commit soon.

yurvyn replied:
Hi @carlescufi

I am testing the new solution (with zephyr-sdk toolchain) and it seems OK, Net Echo Server and Client samples with OpenThread configuration work stable.

However, these is one feature which requires NDS32 toolchain and is not working with zephyr-sdk toolchain. It is CONFIG_FPU (since NDS32 has its own implementation of FPU).

So, we have 2 options:
**Option 1.** We go with NDS32 toolchain as cross-compile and we will ask Andes Technology to add NDS32 toolchain patches to the upstream. 
**Option 2.** We go with zephyr-sdk toolchain and CONFIG_FPU is not supported. In this case we can update the board documentation and add info that to use FPU a custom NDS32 toolchain is required and add the downloadable link to the toolchain.

Cloud you please confirm which option is better for us?

stephanosio replied:
@yurvyn Why not both?

cwshu replied:
@yurvyn Could you describe more detail about `CONFIG_FPU` not working with zephyr-sdk? It looks strange.
D25F supports standard RISC-V F/D extension, so that generic toolchain for RVF/RVD can compile and run programs on D25F. For example, I have compiled and run Zephyr FPU test (`tests/kernel/fpu_sharing`) successfully by Zephyr SDK on D25F on the  FPGA.

yurvyn replied:
@stephanosio,
> Why not both?

The **Option 2** means support of both toolchains. However, CONFIG_FPU is not supported by the toolchain located in zephyr-sdk. For this option, I propose to update the board documentation with the info that in order to use FPU a custom NDS32 toolchain is required. With the **Option 1** CONFIG_FPU is supported by default.

@cwshu,
> Could you describe more detail about `CONFIG_FPU` not working with zephyr-sdk?

I am using three different toolchains: **NDS32 toolchains** (mculib-v5f and mculib-v5 from AndeSight STD v3.2.3) and the generic RISC-V toolchain which is located in zephyr-sdk (**zephyr-sdk toolchain**). I observe the next behaver:
1. **NDS32 Toolchain (mculib-v5f):** I can successfully build Zephyr TLSR9 solution only in case if `CONFIG_FPU=y`. The test `kernel/fpu_sharing` is pass. In case if `CONFIG_FPU=n` it fails to build even simple samples (hello_world, blinky etc..), there is a linker errors: `can't link single-float modules with soft-float modules`.
2. **NDS32 Toolchain  (mculib-v5):** I can successfully build Zephyr TLSR9 solution only in case if `CONFIG_FPU=n`. In case if `CONFIG_FPU=y` it fails to build even simple samples (hello_world, blinky etc..), there is a linker errors: `can't link single-float modules with soft-float modules`.
3. **zephyr-sdk toolchain**: I can build the Zephyr TLSR9 solution (hello_world, blinky etc..) only in case if `CONFIG_FPU=n`. In case if FPU is enabled, there is a linker errors: `can't link soft-float modules with single-float modules`.

@carlescufi, @stephanosio, @galak
So, looks like the right workflow is: I will remove all NDS32 specific instructions and replace them by the generic RISC-V instructions. As result Zephyr TLSR9 solution could be used with the risc-v toolchain from zephyr-sdk. I will also update the board documentation with the info that in order to use FPU a custom specific NDS32 toolchain is required.

yurvyn replied:
Done.
The latest commit supports both zephyr-sdk and cross-compile toolchains.

yurvyn commented:
Hi all,
Next week I will not have access to the internet and I will not be able to response messages.
Will come back on May, 31.

carlescufi commented on west.yml:
```diff
@@ -74,6 +74,9 @@ manifest:
       revision: f8ff8d25aa0a9e65948040c7b47ec67f3fa300df
       path: modules/hal/stm32
+    - name: hal_telink
+      revision:
```
Can you please point this to a revision in a `hal_telink` repo hosted by yourself so we can take a look before creating the repo? You can use the `url` tag to point to a repo outside `zephyrproject-rtos`

yurvyn replied:
Done

carlescufi commented on west.yml:
```diff
@@ -74,6 +74,9 @@ manifest:
     - name: hal_stm32
       revision: f8ff8d25aa0a9e65948040c7b47ec67f3fa300df
       path: modules/hal/stm32
+    - name: hal_telink
+      revision:
+      path: modules/hal/telink
```
```suggestion
    - name: hal_telink
      url: https://github.com/yurvyn/hal_telink
      revision: master
      path: modules/hal/telink
```

yurvyn replied:
Done.
Thanks for the example.

carlescufi reviewed (commented):
I have no objections to merging this (and creating the hal_telink repo), but the toolchain issue needs to be resolved with @galak and @nashif first.

yurvyn commented:
> I have no objections to merging this (and creating the hal_telink repo), but the toolchain issue needs to be resolved with @galak and @nashif first.

@carlescufi,
The toolchain issue is resolved, the zephyr toolchain is supported.
Could you please continue with the PR?

carlescufi commented:
> > I have no objections to merging this (and creating the hal_telink repo), but the toolchain issue needs to be resolved with @galak and @nashif first.
> 
> @carlescufi,
> The toolchain issue is resolved, the zephyr toolchain is supported.
> Could you please continue with the PR?

Hi @yurvyn yes, let us create the actual HAL repo first and the continue.

carlescufi commented:
> The next projects fail to build since NDS32 FPU is not supported by the RISC-V toolchain located in zephyr-sdk:
> 
> * tests/kernel/fpu_sharing/float_disable
> * tests/kernel/fpu_sharing/generic
> * tests/lib/sprintf
> 
> Is it possible to disable these projects in buildkite/zephyr check for tlsr9518adk80d board?

So it's only the FPU that is not currently supported from what I understand.
@galak any input here?

yurvyn commented:
@carlescufi,
> Why not just remove the `select FPU` from the `SOC_RISCV_TELINK_B91` Kconfig option?

It is already removed ([soc: riscv: telink_b91: support of zephyr toolchain](https://github.com/zephyrproject-rtos/zephyr/pull/35533/commits/77f737083d38ee0a09efb0b799af55ac10773f1f)). 
The issue is reproducible only with the `fpu_sharing` and `sprintf` tests project which enable FPU. These projects are pass if use NDS32 toolchain.

The board documentation contains this info: To use Floating Point Unit (``CONFIG_FPU=y``) a Telink specific RISC-V Toolchain is required...

yurvyn commented:
@galak, @carlescufi

The hal_telink module is created and I have just updated the revision in `west.yml`.
If you are done/ok with the review, please merge it.

carlescufi commented:
> The issue is reproducible only with the `fpu_sharing` and `sprintf` test projects which enable FPU. These projects are pass if use NDS32 toolchain.

In that case you will need to exclude the Telink board from those tests.
For the best way to do that I will delegate to @galak, @nashif and @PerMac

yurvyn commented:
@carlescufi,
> maybe you need to disable CPU_HAS_FPU in this case?

I don't think that it is good solution from the interface point of view. With the current solution, to use FPU users just need to enable it via `CONFIG_FPU=y` and use Telink/NDS32 specific toolchain. In case if `CPU_HAS_FPU` is removed, users need to set the both configs, and this is not really good because `CPU_HAS_FPU` is not designed to be set by end users.

yurvyn commented:
@carlescufi,

I have just checked the solution from qemu_riscv32 (the only one RISC-V board with `CPU_HAS_FPU` enabled) and looks like now I can use FPU with the zephyr-sdk toolchain, just need to set `COMPRESSED_ISA` in board's `Kconfig.defconfig` file.
I need some time to finish the testing and tomorrow I will provide a final commit which should pass all the checks.

yurvyn commented:
@carlescufi, @galak,

All CI checks pass.
If you are ok/done with the review, please approve the PR.

ioannisg commented:
> I am not sure that it is good solution from the interface point of view. With the current solution, to use FPU users just need to enable it via CONFIG_FPU=y and use Telink/NDS32 specific toolchain. In case if CPU_HAS_FPU is removed, users need to set the both configs, and this is not really good because CPU_HAS_FPU is not designed to be set by end users.

I agree that CPU_HAS_FPU should signify the HW (SoC) capabilities. So, if the SoC supports it, the option should stay in the SoC definition. If the Zephyr SDK does not support FPU for this particular ARCH/Board,  guess the best option should be to disallow enabling FPU/FPU_SHARING for this case.

carlescufi reviewed (approved):
Approved from the overall structure, module and commit split perspective.

carlescufi reviewed (changes requested):
I was looking at an old history
@yurvyn could you please rebase the commits into a sensible set? One for the SoC, one for the board, one for each driver.

carlescufi reviewed (approved):
Thanks, the commit set looks sane now.

yurvyn commented:
@galak,
Cold you please approve the PR?

mbolivar-nordic commented on dts/bindings/cpu/riscv,b91.yaml:
```diff
@@ -0,0 +1,8 @@
+
+description: Telink RISC-V CPU
+
+compatible: "riscv,b91"
```
Looking here: https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/riscv/cpus.yaml

I think it would be more appropriate to say "telink,b91" here.

You can still say `compatible = "telink,b91", "riscv";` in a DTSI file as done in the SiFive example in the above link.

yurvyn replied:
Done.

mbolivar-nordic replied:
Thanks for your quick response, but it looks like this change was done in an additional commit in the same PR. Zephyr policy is that changes in response to review should be squashed into the original commits, not put on top. Please rebase the patch series so requested changes are not handled as additional commits, but squashed appropriately.

yurvyn replied:
Done.

mbolivar-nordic commented on boards/riscv/tlsr9518adk80d/Kconfig.defconfig:
```diff
@@ -0,0 +1,17 @@
+# Copyright (c) 2021 Telink Semiconductor
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD
```
It looks like the whole file should be inside an `if BOARD_TLSR9518ADK80D` instead.

yurvyn replied:
Done

mbolivar-nordic commented on boards/riscv/tlsr9518adk80d/pinmux.c:
```diff
@@ -0,0 +1,85 @@
+{
+	ARG_UNUSED(dev);
+
+	const struct device *pinmux =
+		device_get_binding(DT_LABEL(DT_NODELABEL(pinmux)));
```
No need for device_get_binding or label properties anymore.

```c
const struct device *pinmux = DEVICE_DT_GET(DT_NODELABEL(pinmux));
```

If it is possible that the pinmux device may fail to initialize, this must be followed with:

```c
if (!device_is_ready(pinmux)) {
        return -ENODEV;
}
```

yurvyn replied:
Fixed. Please pay attention that pins configuration is moved to DTS (see Serial driver).

mbolivar-nordic commented on boards/riscv/tlsr9518adk80d/tlsr9518adk80d.dts:
```diff
@@ -0,0 +1,124 @@
+
+/ {
+	model = "telink,b91";
+	compatible = "telink,b91";
```
"telink,tlsr9518adk80d"

yurvyn replied:
Done.

mbolivar-nordic commented on drivers/gpio/gpio_b91.c:
```diff
@@ -0,0 +1,542 @@
+
+
+/* Initialize interrupt handler for IRQ */
+static void irq_connect(void)
```
This is called from gpio_b91_init(), which means it gets run for every single instance, connecting the same IRQ multiple times. I think it would be better to provide a per-instance initialization function that connects the IRQ for that instance. See gpio_mcux.c for an example.

yurvyn replied:
Done. Now each instance has its own connect function.
Please pay attention that ``gpio_b91_irq_connect_0..4()`` is not moved to ``DT_NUM_INST_STATUS_OKAY()`` because it is not possible to use ``#if IS_INST_IRQ_EN(n)`` inside ``DT_NUM_INST_STATUS_OKAY(..)``. We should not connect IRQ if it is not selected in DTS.

mbolivar-nordic replied:
> We have to use `#if IS_INST_IRQ_EN(n)` because we should not connect IRQ if it is not selected in DTS.

I'm confused about this.

I see:

```c
+/* Check that 'inst' has only 1 interrupt selected in dts */
+#define IS_INST_IRQ_EN(inst)    (DT_INST_IRQ_HAS_IDX(inst, 0) != 0) && \
+	(DT_INST_IRQ_HAS_IDX(inst, 1) == 0)
```

But the DTS files mention 3 interrupts, it looks like:

```
		gpioa: gpio@80140300 {
...
			interrupt-parent = <&plic0>;
			interrupts = <25 1>, <26 1>, <27 1>;
...
		};
		plic0: interrupt-controller@e4000000 {
			compatible = "sifive,plic-1.0.0";
			#interrupt-cells = <2>;
			interrupt-controller;
...
		};

```

What's going on here?

yurvyn replied:
It works in the next way:

Each GPIO node in **DTSI** (soc-level) file has list of all supported IRQs. It informs users about port capabilities, which IRQ numbers are supported by the port.

```
gpioa: gpio@80140300 {
...
interrupts = <25 1>, <26 1>, <27 1>;
...
};
```
In **DTS** (board-level) file, if user wants to use GPIO port with interrupt, user should select **only one IRQ**:
```
&gpioc {
	interrupts = <25 1>;
	status = "okay";
};
```
On the driver side: driver checks if `interrupts` has **only one node**. It means that user has selected it in the Board's DTS file and really wants to use/config interrupts.

We have such specific implementation for GPIO driver because of the next **silicon limitation**:
We have 5 GPIO ports (port A .. port E) but only 3 IRQ numbers (25 .. 26) which could be assigned for the port.
It means that maximum 3 out of 5 GPIO ports cold be configured to detect interrupts and the decision should be done by user in board-level **DTS** file. It means that we cannot call IRQ_CONNECT for each instance of GPIO driver, we are doing it only for instances which are selected by users in board-level **DTS** file.

@mbolivar-nordic,
For me the current solution looks quite user-friendly and currently I do not see any other implementation options which are more user-friendly.
In case if you have any other ideas please inform me.

mbolivar-nordic replied:
> On the driver side: driver checks if interrupts has only one node. It means that user has selected it in the Board's DTS file and really wants to use/config interrupts. [...]
>  In case if you have any other ideas please inform me.

Sorry for the slow response; I was on vacation. Can you please explain why checking that `DT_NUM_IRQS(...) == 1` is not what you want?

RIght now, the check is that index 0 exists but index 1 does not. That seems weird given that there are 3 possible indexes.

yurvyn replied:
Yes, you are right. Somehow I haven't seen `DT_NUM_IRQS`.
Fixed.
```
#define IS_INST_IRQ_EN(inst)    (DT_NUM_IRQS(DT_DRV_INST(inst)) == 1)
```

mbolivar-nordic commented on drivers/serial/uart_b91.c:
```diff
@@ -0,0 +1,571 @@
+			      CONFIG_KERNEL_INIT_PRIORITY_DEVICE,		    \
+			      (void *)&uart_b91_driver_api);			    \
+										    \
+	static void uart_b91_irq_connect_##n(void)				    \
```
Looks like you are already familiar with the technique :)

yurvyn replied:
Yes :) 
It was not implemented for GPIO because I had issues with ``#if IS_INST_IRQ_EN(n)`` usage inside ``DT_NUM_INST_STATUS_OKAY(..)``. But now, looks like, I have provided good solution for GPIO.

mbolivar-nordic reviewed (changes requested):
Thanks for the nicely organized DTS related patches. This was really easy to review. I have a few minor questions and nitpicks related to DT usage.

mbolivar-nordic commented on boards/riscv/tlsr9518adk80d/pinmux.c:
```diff
@@ -0,0 +1,85 @@
+	reg_gpio_pad_mul_sel |= BIT(0);
+
+	/* UART0 RX, TX */
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(uart0), okay) && CONFIG_UART_TELINK_B91
+	pinmux_pin_set(pinmux, UART0_TX_PB2, UART0_TX_PB2_FUNC);
+	pinmux_pin_set(pinmux, UART0_RX_PB3, UART0_RX_PB3_FUNC);
+#endif
```
Not sure, but @galak may want this in DTS instead.

yurvyn replied:
Yes, @galak already asked me to move it to dts during an another PR.
Done.

stephanosio reviewed (commented):
IIUC, B91 is the name of the development board (EVK), and the actual SoC family is called TLSR9.

If that is the case, the B91 nomenclature should only be used by the board portion. The rest (SoC and drivers) should use the TLSR9 nomenclature.

yurvyn commented:
@stephanosio,

> IIUC, B91 is the name of the development board (EVK), and the actual SoC family is called TLSR9.
> 
> If that is the case, the B91 nomenclature should only be used by the board portion. The rest (SoC and drivers) should use the TLSR9 nomenclature.

B91 definitely is related to SoC name, it is like alias for TLSR9.
Below you can see that all the rest SDKs related to this family also start with "B91":
http://wiki.telink-semi.cn/tools_and_sdk/BLE/B91_BLE_SDK.zip
http://wiki.telink-semi.cn/tools_and_sdk/BLE_Mesh/SIG_Mesh/B91_SIG_Mesh_SDK.zip
http://wiki.telink-semi.cn/tools_and_sdk/Driver/B91_Driver_SDK.zip

However, I have to discuss the renaming of "B91" to "TLSR9" in Zephyr with my colleges.
I will come back once I have updates.

Thanks, good point.

yurvyn commented:
> > IIUC, B91 is the name of the development board (EVK), and the actual SoC family is called TLSR9.
> > If that is the case, the B91 nomenclature should only be used by the board portion. The rest (SoC and drivers) should use the TLSR9 nomenclature.
> 
> B91 definitely is related to SoC name, it is like alias for TLSR9.
> Below you can see that all the rest SDKs related to this family also start with "B91":
> http://wiki.telink-semi.cn/tools_and_sdk/BLE/B91_BLE_SDK.zip
> http://wiki.telink-semi.cn/tools_and_sdk/BLE_Mesh/SIG_Mesh/B91_SIG_Mesh_SDK.zip
> http://wiki.telink-semi.cn/tools_and_sdk/Driver/B91_Driver_SDK.zip
> 
> However, I have to discuss the renaming of "B91" to "TLSR9" in Zephyr with my colleagues.
> I will come back once I have updates.
> 
> Thanks, good point.

@stephanosio,

I have just discussed the SoC name with the colleagues once more (the first time I discussed it during the development) and the decision is to go with B91.

B91 is an alias for TLSR9 and means:
B - BT/BLE/Zigbee/2.4G
9 - RISC-V
1 -  product generation

On the Telink website both family names are used TLSR9 and B91.
But, the reason why we should go with B91 is that we want to be consistent with all the rest SDKs which are using B91 only.

yurvyn commented:
@mbolivar-nordic, @galak, @stephanosio, @ioannisg 

If you are done/ok with the review, could you please approve it.

yurvyn commented:
@mbolivar-nordic, @galak, @stephanosio, @ioannisg

Are you OK with the PR? please approve it.

stephanosio commented on drivers/gpio/gpio_b91.c:
```diff
@@ -0,0 +1,564 @@
+	/* GPIO function enable */
+	WRITE_BIT(gpio->actas_gpio, BIT(pin), 1);
+
+	/* Setting Drive Strength is not supported by Zephyr */
```
Actually, they are supported:
https://github.com/zephyrproject-rtos/zephyr/blob/be2e6a08500ef8c157639b89d27c0a858ec43dcb/include/drivers/gpio.h#L212

yurvyn replied:
Fixed. 
Removed this comment and added verification for drive strength. Currently we are not supporting it

stephanosio commented on drivers/pinmux/pinmux_b91.c:
```diff
@@ -0,0 +1,245 @@
+ *      gpio_en + 4*8:    PORT_E[0-7]
+ *      gpio_en + 5*8:    PORT_F[0-7]
+ */
+#define reg_gpio_en(pin) (*(volatile uint8_t *)((uint32_t)DT_INST_REG_ADDR_BY_NAME(0, gpio_en) + \
+						((pin >> 8) << 3)))
```
Nit: Unless `<< 3` is semantically a bitfield operation, `* 8` should be preferred as that makes it clearer as to what is being done.
```suggestion
#define reg_gpio_en(pin) (*(volatile uint8_t *)((uint32_t)DT_INST_REG_ADDR_BY_NAME(0, gpio_en) + \
						((pin >> 8) * 8)))
```

yurvyn replied:
Fixed in all places.

stephanosio commented on drivers/serial/uart_b91.c:
```diff
@@ -0,0 +1,588 @@
+
+/* B91 UART registers structure */
+struct uart_b91_t {
+	uint8_t data_buf0;
+	uint8_t data_buf1;
+	uint8_t data_buf2;
+	uint8_t data_buf3;
```
Any reason we can't do `uint8_t data_buf[4]` here?

yurvyn replied:
Fixed, uint8_t data_buf[4] is used.

stephanosio commented on drivers/serial/uart_b91.c:
```diff
@@ -0,0 +1,588 @@
+	while (uart_b91_get_tx_bufcnt(uart) >= UART_TX_BUF_CNT) {
+	};
+
+	if (data->tx_byte_index == 0) {
+		uart->data_buf0 = c;
+	} else if (data->tx_byte_index == 1) {
+		uart->data_buf1 = c;
+	} else if (data->tx_byte_index == 2) {
+		uart->data_buf2 = c;
+	} else if (data->tx_byte_index == 3) {
+		uart->data_buf3 = c;
+	}
```
Do we really have to do this?

yurvyn replied:
Fixed, uint8_t data_buf[4] is used. Removed all if..else..

stephanosio commented on drivers/serial/uart_b91.c:
```diff
@@ -0,0 +1,588 @@
+		uart->data_buf3 = c;
+	}
+
+	data->tx_byte_index++;
+	data->tx_byte_index &= 0x03;
```
We are well past the 80s when C compilers were dumb.
```suggestion
	data->tx_byte_index = (data->tx_byte_index + 1) % 4;
```
OR, even better:
```suggestion
	data->tx_byte_index = (data->tx_byte_index + 1) % ARRAY_SIZE(uart->data_buf);
```

yurvyn replied:
Done.

stephanosio commented on drivers/serial/uart_b91.c:
```diff
@@ -0,0 +1,588 @@
+			      const uint8_t *tx_data,
+			      int size)
+{
+	int i = 0;
```
Redundant initialisation

yurvyn replied:
Fixed.

stephanosio commented on drivers/serial/uart_b91.c:
```diff
@@ -0,0 +1,588 @@
+	volatile struct uart_b91_t *uart = GET_UART(dev);
+
+	uart->ctrl3 = (uart->ctrl3 & (~FLD_UART_TX_IRQ_TRIQ_LEV)) |
+		      (0x01 << FLD_UART_TX_IRQ_TRIQ_LEV_OFFSET);
```
Why not `BIT(FLD_UART_TX_IRQ_TRIQ_LEV_OFFSET)`?

yurvyn replied:
Done.

stephanosio reviewed (changes requested):
Some comments

galak commented on boards/riscv/tlsr9518adk80d/pin_func.h:
```diff
@@ -0,0 +1,117 @@
+/**
+ *  @brief  Define UART0 pins : TX(PA3 PB2 PD2), RX(PA4 PB3 PD3)
+ */
+#define UART0_TX_PA3            GPIO_PA3
```
We should get these settings defined in devicetree.  See how STM32 or ATMEL SoCs are doing this.

yurvyn replied:
Done. Implemented the same solution as STM32 has.
Now it is much better.. thanks.

stephanosio commented on soc/riscv/riscv-privilege/telink_b91/soc.c:
```diff
@@ -0,0 +1,144 @@
+}
+
+#if defined(CONFIG_NEWLIB_LIBC)
+/**
+ * @brief Implemented missed __ashldi3 in case if newlib is selected.
+ */
+long __ashldi3(long a, int b)
+{
+	return a << b;
+}
+
+/**
+ * @brief Implemented missed __lshrdi3 in case if newlib is selected.
+ */
+long __lshrdi3(long a, int b)
+{
+	return a >> b;
+}
```
A little curious as to why this is necessary because these functions are supposed to be implemented by the libgcc.

Do we need this when building with the Zephyr SDK as well? or just for Telink's toolchain?

yurvyn replied:
Hm.. Now it is working even with Telink toolchain.
Probably something is changed in Zephyr build system on the lattes commits.

Done, `__ashldi3` and `__lshrdi3` are removed. 
Good catch, thanks.

mbolivar-nordic commented:
> If you are done/ok with the review, could you please approve it

Not sure about the GPIO interrupt handling here.

yurvyn commented on boards/riscv/tlsr9518adk80d/pin_func.h:
```diff
@@ -0,0 +1,117 @@
+ */
+
+#ifndef __TELINK_TLSR9518ADK80D_PIN_FUNC_H_
+#define __TELINK_TLSR9518ADK80D_PIN_FUNC_H_
```
Fixed. `Pin_func.h` is removed in scope of an another fix.

yurvyn commented on boards/riscv/tlsr9518adk80d/pin_func.h:
```diff
@@ -0,0 +1,117 @@
+#ifndef __TELINK_TLSR9518ADK80D_PIN_FUNC_H_
+#define __TELINK_TLSR9518ADK80D_PIN_FUNC_H_
+
+#define BIT(n)                 (1<<(n))
```
Fixed. `Pin_func.h` is removed in scope of an another fix.

yurvyn commented on drivers/gpio/gpio_b91.c:
```diff
@@ -0,0 +1,564 @@
+#define reg_irq_risc1_en(i) REG_ADDR8(0x140340 + i)
+
+/* Pull-up/down resistors */
+#define GPIO_PIN_UP_DOWN_FLOAT   ((uint8_t) 0u)
```
Fixed.

yurvyn commented on drivers/gpio/gpio_b91.c:
```diff
@@ -0,0 +1,564 @@
+		BM_SET(reg_irq_risc0_en(GET_PORT_NUM(gpio)), BIT(pin));
+	} else if (irq == IRQ_GPIO2_RISC1) {
+		BM_SET(reg_irq_risc1_en(GET_PORT_NUM(gpio)), BIT(pin));
+	}
```
Done.

yurvyn commented:
@galak, @stephanosio, @mbolivar-nordic

All your comments are fixed.
Please review.

yurvyn commented:
@galak, @stephanosio, @mbolivar-nordic

`West setup` check fails with ERROR: update failed for project tinycrypt.
Looks like it is not related to my changes.

Should I do something with it?

yurvyn commented:
>> @galak, @stephanosio, @mbolivar-nordic
>> 
>> All your comments are fixed.
>> Please review.

>@galak, @stephanosio, @mbolivar-nordic
>
>West setup check fails with ERROR: update failed for project tinycrypt.
>Looks like it is not related to my changes.
>
>Should I do something with it?

@galak, @stephanosio, @mbolivar-nordic

In case if you are OK with my fixes and if you do not have new comments, please approve the review.

yurvyn commented:
@galak, @stephanosio, @mbolivar-nordic

I already have a list of other drivers/subsystems which are waiting for PR, such as: IEEF802154/OpenThread, PWM, Flash, Entropy, SPI, I2C etc. Pull requests for such divers can be created in parallel, but they all are blocked by this basic PR.

**Could you please speed up the review?**

Commit:
boards: riscv: tlsr9518adk80d: new Telink B91 board support

New Telink RISC-V B91 Generic Starter Kit board support.

Signed-off-by: Yuriy Vynnychek <yura.vynnychek@telink-semi.com>

Commit:
soc: riscv: telink_b91: new Telink B91 (TLSR9) SoC system

A new RISC-V platform SoC for Telink B91 (TLSR9).

Signed-off-by: Yuriy Vynnychek <yura.vynnychek@telink-semi.com>

Commit:
include: dt-bindings: b91_pinctrl: add pinctrl bindings for B91

Added DT pinctrl bindings for new Telink B91 SoC.

Signed-off-by: Yuriy Vynnychek <yura.vynnychek@telink-semi.com>

Commit:
dts: riscv: add basic dts support for new Telink B91 platform

CPU, Power, GPIO, Serial and Pinctrl dtsi support for new
Telink B91 RISC-V platform.

Signed-off-by: Yuriy Vynnychek <yura.vynnychek@telink-semi.com>

Commit:
modules: hal_telink: introduce new HAL Telink module

This commit introduce new HAL Telink module used by
new Telink B91 RISC-V platform drivers.

Signed-off-by: Yuriy Vynnychek <yura.vynnychek@telink-semi.com>

Commit:
drivers: gpio: introduce new Telink B91 GPIO driver

GPIO driver basic support for new Telink B91 platform.

Signed-off-by: Yuriy Vynnychek <yura.vynnychek@telink-semi.com>

Commit:
drivers: pinmux: introduce new Telink B91 Pinmux driver

Pinmux driver basic support for new Telink B91 platform.

Signed-off-by: Yuriy Vynnychek <yura.vynnychek@telink-semi.com>

Commit:
drivers: serial: introduce new Telink B91 Serial driver

Serial driver basic support for new Telink B91 platform.

Signed-off-by: Yuriy Vynnychek <yura.vynnychek@telink-semi.com>

yurvyn commented:
Hello @galak, @stephanosio, @mbolivar-nordic 

Looks like you have reviewed all files.
From my side all comments are fixed, CI checks are passed.

Could you please approve the PR?

stephanosio commented on drivers/serial/uart_b91.c:
```diff
@@ -0,0 +1,568 @@
+/* B91 UART registers structure */
+struct uart_b91_t {
+	uint8_t data_buf[4];
+	uint8_t clk_div_l;
+	uint8_t clk_div_h;
```
I assume B91 is little endian and there will never be a big endian variant. In that case, any reason we can't do `uint16_t clk_div_l` here? Do these peripherals have only lower 8 bits connected to the data bus?

yurvyn replied:
16 bits are connected.
Fixed.

stephanosio commented on drivers/serial/uart_b91.c:
```diff
@@ -0,0 +1,568 @@
+	uint8_t ctrl1;
+	uint8_t ctrl2;
+	uint8_t ctrl3;
+	uint8_t rxtimeout_l;
+	uint8_t rxtimeout_h;
```
Same here.

yurvyn replied:
Done.

stephanosio commented on drivers/serial/uart_b91.c:
```diff
@@ -0,0 +1,568 @@
+#else
+	struct uart_b91_data *data = DEV_DATA(dev);
+
+	if (data->callback) {
```
```suggestion
	if (data->callback != NULL) {
```

yurvyn replied:
Done.

stephanosio commented on drivers/serial/uart_b91.c:
```diff
@@ -0,0 +1,568 @@
+}
+
+/* API implementation: poll_in */
+static int uart_b91_poll_in(const struct device *dev, unsigned char *c)
+{
+	volatile struct uart_b91_t *uart = GET_UART(dev);
+	struct uart_b91_data *data = DEV_DATA(dev);
+
+	*c = uart->data_buf[data->rx_byte_index];
+	data->rx_byte_index = (data->rx_byte_index + 1) % ARRAY_SIZE(uart->data_buf);
+
+	return 0;
+}
```
This implementation does not look sane.

`poll_in` is supposed to return data if there is data available in the RX buffer. This implementation is missing any such checks.

https://github.com/zephyrproject-rtos/zephyr/blob/622a8e825658e91e4afed8d6986cc97b81229484/include/drivers/uart.h#L661-L663

yurvyn replied:
Fixed.

	if (uart_b91_get_rx_bufcnt(uart) == 0) {
		return -1;
	}

stephanosio commented on drivers/serial/uart_b91.c:
```diff
@@ -0,0 +1,568 @@
+	while (uart_b91_get_tx_bufcnt(uart) >= UART_TX_BUF_CNT) {
+	};
+
+	uart->data_buf[data->tx_byte_index] = c;
+	data->tx_byte_index = (data->tx_byte_index + 1) % ARRAY_SIZE(uart->data_buf);
```
This UART driver implementation seems to share the same data buffer for TX and RX data. Does this UART peripheral not support duplex operation?

yurvyn replied:
Yes, TX and RX are using the same HW buffer. This is how the Silicon is implemented.

stephanosio commented on soc/riscv/riscv-privilege/telink_b91/start.S:
```diff
@@ -0,0 +1,86 @@
+
+start:
+
+#if 0
+	/* Debug LED */
+	lui t0,0x80140
+	li t1, 0x0f
+	li t2, 0x10
+	sb t1 , 0x30a(t0)
+	sb t2 , 0x30b(t0)
+#endif
```
Did you forget to remove this?

yurvyn replied:
Yes, removed.

stephanosio commented on soc/riscv/riscv-privilege/telink_b91/start.S:
```diff
@@ -0,0 +1,86 @@
+	sb t2 , 0x30b(t0)
+#endif
+
+#ifdef __nds_execit
+	/* Initialize EXECIT table */
+	la t0, _ITB_BASE_
+	csrw uitb, t0
+#endif
```
How is `__nds_execit` defined here?

yurvyn replied:
NDS toolchain specific stuff.
We do not need it now. Removed.

stephanosio reviewed (dismissed):
More comments

yurvyn commented on soc/riscv/riscv-privilege/telink_b91/soc.h:
```diff
@@ -0,0 +1,17 @@
+ */
+
+#ifndef __RISCV_TELINK_B91_SOC_H_
+#define __RISCV_TELINK_B91_SOC_H_
```
Fixed

yurvyn commented on drivers/pinmux/pinmux_b91.c:
```diff
@@ -0,0 +1,245 @@
+	if (status != 0) {
+		return status;
+	}
+	mask = (uint8_t) ~(BIT(offset) | BIT(offset + 1));
```
Fixed

yurvyn commented on boards/riscv/tlsr9518adk80d/doc/index.rst:
```diff
@@ -0,0 +1,208 @@
+
+.. code-block:: console
+
+   # Set Zephyr toolchain varian to cross-compile
```
Fixed

yurvyn commented on boards/riscv/tlsr9518adk80d/tlsr9518adk80d.dts:
```diff
@@ -0,0 +1,125 @@
+			label = "image-scratch";
+			reg = <0x0003c000 0x2000>;
+		};
+		storage_partition: partition@F0000 {
```
Fix.
We are not getting any dtc warnings about this.

yurvyn commented on boards/riscv/tlsr9518adk80d/tlsr9518adk80d.dts:
```diff
@@ -0,0 +1,125 @@
+		};
+		storage_partition: partition@F0000 {
+			label = "storage";
+			reg = <0x000F0000 0x00010000>;
```
Done.

yurvyn commented:
@galak, @stephanosio, @mbolivar-nordic

All comments are fixed, CI checks are passed.
Please approve the PR.

mbolivar-nordic commented:
> All comments are fixed, CI checks are passed.

One of my comments was marked resolved but I'm not sure of the response, so I unresolved it.

yurvyn commented:
> One of my comments was marked resolved but I'm not sure of the response, so I unresolved it.

@mbolivar-nordic,

Please check the fix.
If everything is OK and and if you are done with the review, cloud you please approve the PR?

mbolivar-nordic reviewed (approved):
Thanks for your patience and the IRQ change! My comments are addressed.

yurvyn commented:
Hello @stephanosio,

The last one change request is left which is blocking merging. 

Could you please check my fixes and mark them as resolved?

stephanosio reviewed (commented):
I haven't had time to go through the entire series again, but the previous comments seem to have been addressed and it looks ok overall.

yurvyn commented:
@carlescufi, @stephanosio, @galak, @ioannisg, @mbolivar-nordic 

Could anyone please merge the PR?
